### DISCUSSION TITLE
How to trigger node transforms when importing editor state (parseEditorState/setEditorState)

### DISCUSSION THREAD

**Discussion by birtles** [facebook/lexical#3055] [Q&A]
I have a node type that generates additional helper nodes as part of a node transform.

However, in some circumstances I am updating the lexical state using `parseEditorState` to create a new `EditorState` and then setting it using `setEditorState`.

When I perform these steps, the node transforms are not run and hence the helper nodes are not generated.

I can't force these helper nodes to be generated in `importJSON` since it relies on the node's children being present and `importJSON` never sees the children for the node (unlike `importDOM` which can provide an `after` callback for each conversion).

I can force the node transforms to be run by borrowing the relevant bit of code from `markAllNodesAsDirty`:

```ts
editor.setEditorState(newEditorState);
editor.update(() => {
  const nodeMap = editor.getEditorState()._nodeMap;
  for (const [, node] of nodeMap) {
    node.markDirty();
  }
});
```

However, it seems fragile to be reaching into `_nodeMap` like this.

How should this work?

- Should there be a primitive for forcing all node transforms to run?
- Should `setEditorState` do it automatically?
- Should there be a callback for `importDOM` to update the node after its children have been attached?

Any ideas?

**Comment by acywatson** [comment]
Interesting ideas - looks like we have another issue around this too: https://github.com/facebook/lexical/issues/3200

**Comment by sebakerckhof** [comment]
Ah, yes, I actually ended up with the same hacky workaround. But I figured it was a bug since it seemed logical to me that tranforms would also be called on setEditorState. Otherwise nodes just 'magically'/suddenly transform when they're marked as dirty later on. On the other hand, a primitive to force node transforms to run may be more flexible.
